ring/client/pool.go

```diff
@@ -173,9 +175,18 @@ func (p *Pool) removeStaleClients() {
 	}
 }
 
-// cleanUnhealthy loops through all servers and deletes any that fails a healthcheck.
+// cleanUnhealthy loops through all servers and deletes any that fails a healthcheck
+// The health checks are executed concurrently with p.cfg.MaxConcurrentHealthChecks.
+// If this is not set (0), then all healthchecks of all clients are executed in parallel.
 func (p *Pool) cleanUnhealthy() {
-	for _, addr := range p.RegisteredAddresses() {
+	addresses := p.RegisteredAddresses()
+
+	maxConcurrent := p.cfg.MaxConcurrentHealthChecks
+	if p.cfg.MaxConcurrentHealthChecks == 0 {
+		maxConcurrent = len(addresses)
+	}
+	_ = concurrency.ForEachJob(context.Background(), len(addresses), maxConcurrent, func(_ context.Context, idx int) error {
```

Comment:
If we're executing these checks concurrently, I wonder if we need to cancel the checks if they overrun the interval at which `cleanUnhealthy` is invoked - otherwise they'll start to run on divergent intervals. Is that even an issue?

Reply:
Create a context with timeout instead of passing `context.Background()` ?

Reply:
Which timeout should we pass?

Reply:
Maybe `CheckInterval`?

Reply:
So, when I pass a cancellable context or a context with timeout to the `healtCheck()` function and the context gets cancelled, it could be that not all health checks have been executed - which is ok. However, there is a race condition, where the context is cancelled and there are health check requests currently in flight. In that case, a context cancellation would cause the request to fail and the client to be removed, even though it may have been healthy, right?

Reply:
@bboreham @pracucci Since all child contexts have a timeout, do I really need to set a timeout on the parent context as well? The only case I could think of is when the health check timeout is greater than the check interval - which should never be the case and we could also add an assertion for that.

Reply:
Forget what I said. The function `concurrency.ForEachJob()` is waiting for all health checks to finish :man_facepalming: We don't need to set a timeout on the parent context.